Add remove_duplicates.py to Ch.11

srikar-eranky opened the pull request:
Srikar Eranky

**Remove_duplicates.py**
* Chapter 11
* Practice



Added a couple hints to the practice template for sorting and counting.

Commit:
Add Python problem and solution for Ch.11

lint-action committed:
Fix code style issues with Black

phrdang committed:
Merge branch 'master' into ch11-exercises

Citrus716 reviewed (changes requested):
Nice Problem, but I don't think the solution works all the time. I tried the list: [2, 2, 1, 2, 324, 52, 34, 324], and I got [1, 2, 2, 34, 52, 324, 324]. Let me know if you need any help.

abhatia1205 reviewed (changes requested):
Hi Srikar.

Generally, it is bad to change a list while you are iterating over it. It messes up the loop and often leads to errors, which has happened in your case. For example, if I try the list [1,1,1,1,1,1,1], it returns [1,1], which is incorrect. I would suggest changing the code such that you are not changing the list while iterating. Having a worse space complexity by making a new list is fine because iterating and changing the list at the same time is usually bad practice.

Another note: You should make sure to specify that using the set function is NOT ALLOWED. If it was allowed, the function would just be one line of code: return list(set(array)).

Please make these changes and resubmit. Thanks!

abhatia1205 commented on 2_intermediate/chapter11/practice/remove_duplicates.py:
```diff
@@ -0,0 +1,13 @@
+# The sole parameter of the function should be a list
+# The function should look through a list,
+# Find all duplicate elements, and remove them
+# Sort the resulting list
```
```suggestion
# Sort the resulting list
# YOU MAY NOT USE THE set() function IN PYTHON. Use this problem to practice list iteration!
```

abhatia1205 commented on 2_intermediate/chapter11/solutions/remove_duplicates.py:
```diff
@@ -0,0 +1,31 @@
+# The sole parameter of the function should be a list
+# The function should look through a list,
+# Find all duplicate elements, and remove them
+# Sort the resulting list
```
```suggestion
# Sort the resulting list
# YOU MAY NOT USE THE set() function IN PYTHON. Use this problem to practice list iteration!
```

abhatia1205 commented on 2_intermediate/chapter11/solutions/remove_duplicates.py:
```diff
@@ -0,0 +1,31 @@
+
+
+def remove_duplicate(array):
+    for i in array:
```
It's possible, even probable, that Python gets all the elements in ```array``` only once, which is why when you call ```array.remove()```, ```array``` changes but what the loop is iterating over does not, which causes your solution to crash.

srikar-eranky replied:
Hmm...I'll make a new solution and check if it works, and then I'll fix it. Thanks for the input.

srikar-eranky committed:
Update 2_intermediate/chapter11/practice/remove_duplicates.py

Co-authored-by: abhatia1205 <32373316+abhatia1205@users.noreply.github.com>

srikar-eranky committed:
Merge branch 'master' into ch11-exercises

lint-action committed:
Fix code style issues with Black

srikar-eranky committed:
Update 2_intermediate/chapter11/solutions/remove_duplicates.py

Co-authored-by: abhatia1205 <32373316+abhatia1205@users.noreply.github.com>

srikar-eranky committed:
Update remove_duplicates.py

Redid the solution using dictionaries, didn't realize that they were covered in Ch.9. If you plan on teaching Ch.11 before Ch.9, highly recommend that you don't use this problem.

srikar-eranky committed:
Update remove_duplicates.py

abhatia1205 commented:
Ill just go ahead and merge this. Formatting seems ok.
